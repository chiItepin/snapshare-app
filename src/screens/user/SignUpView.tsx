import React, { FunctionComponent, useState } from 'react';
import moment from 'moment';
import {
  Text,
  ScrollView,
} from 'react-native';
import {
  Button,
  Incubator,
  Toast,
  View,
} from 'react-native-ui-lib';
import { NavigationProp } from '@react-navigation/native';
import { AxiosResponse } from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { connect } from 'react-redux';
import Redux from 'redux';
import IUser from '../templates/user';
import useApi from '../../useApi';
import styles from '../../styles/GlobalStyles';
import typographyStyles from '../../styles/TypographyStyles';
import helperStyles from '../../styles/HelperStyles';
import RootScreenParams from '../RootScreenParams';

const { TextField } = Incubator;

type TSetUser = (userModel: IUser) => void

interface DispatchProps {
  setUser: TSetUser
}
interface IPropsNavigation {
  navigation: NavigationProp<RootScreenParams>;
  setUser: TSetUser
}

const Login:FunctionComponent<IPropsNavigation> = ({
  navigation,
  setUser,
}: IPropsNavigation) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [loading, setLoading] = useState(false);
  const [notificationMessage, setNotificationMessage] = useState('');
  const { User } = useApi();

  const handleSignup = (): void => {
    setLoading(true);
    setNotificationMessage('Loading...');
    User.createUser(email, password, confirmPassword)
      .then(async (res: AxiosResponse) => {
        setLoading(false);
        setNotificationMessage('Signed Up successfully!');

        const newUser: IUser = res.data.data;

        const userModel = {
          email: newUser.email,
          _id: newUser._id,
          token: res.data.token,
          loggedInDate: moment().format('LL'),
        } as IUser;

        await AsyncStorage.setItem('@user', JSON.stringify(userModel));
        setUser(userModel);
      })
      .catch(() => {
        setNotificationMessage('Email already found!');
        setLoading(false);
      });
  };

  return (
    <ScrollView contentContainerStyle={{ flexGrow: 1 }}>
      <View style={styles.container}>
        <Toast
          visible={!!notificationMessage}
          position="bottom"
          showDismiss
          onDismiss={() => setNotificationMessage('')}
          backgroundColor="black"
          autoDismiss={3000}
          message={notificationMessage}
        />

        <View style={styles.card}>
          <Text
            style={[typographyStyles.h1, helperStyles.marginVerticalMed]}
          >
            Signing Up
          </Text>

          <TextField
            placeholder="Enter your email"
            label="Email"
            value={email}
            onChangeText={setEmail}
            autoCapitalize="none"
            enableErrors
            validationMessage={['Email is required', 'Email is invalid']}
            validate={['required', 'email']}
            validateOnChange
            fieldStyle={styles.genericField}
            validationMessageStyle={{ marginBottom: 10 }}
          />

          <TextField
            placeholder="Enter your password"
            label="Password"
            value={password}
            onChangeText={setPassword}
            secureTextEntry
            maxLength={20}
            showCharCounter
            fieldStyle={styles.genericField}
          />

          <TextField
            placeholder="Confirm your password"
            label="Confirm Password"
            value={confirmPassword}
            onChangeText={setConfirmPassword}
            secureTextEntry
            maxLength={20}
            showCharCounter
            fieldStyle={styles.genericField}
          />

          <Button
            onPress={handleSignup}
            label="Sign Up"
            disabled={!password || !email || loading || (password !== confirmPassword)}
            enableShadow
            borderRadius={6}
          />

          <Button
            onPress={() => navigation.navigate('Login')}
            hyperlink
            text90
            link
            label="Already have an account"
            style={helperStyles.marginTopMed}
            labelStyle={{ textDecorationLine: 'none' }}
          />
        </View>
      </View>
    </ScrollView>
  );
};

const mapDispatchToProps = (dispatch: Redux.Dispatch<any>) => ({
  setUser: (userModel: IUser) => dispatch({
    type: 'SET_USER',
    payload: {
      ...userModel,
    },
  }),
});

export default connect<DispatchProps>(null, mapDispatchToProps)(Login);
