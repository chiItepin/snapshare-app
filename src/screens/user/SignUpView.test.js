import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import SignUpView from './SignUpView';
import AppReducer from '../../reducer';

test('SignUpView renders and handles events', async () => {
  const store = createStore(AppReducer);

  const { getByPlaceholderText, getByText, toJSON } = await render(
    <Provider store={store}>
      <SignUpView />
    </Provider>,
  );

  const emailField = getByPlaceholderText('Enter your email');
  const passwordField = getByPlaceholderText('Enter your password');
  const confirmPassword = getByPlaceholderText('Confirm your password');
  const submitButton = getByText('Sign Up');

  await act(async () => {
    fireEvent.changeText(emailField, 'lorem@ipsum.com');
    fireEvent.changeText(passwordField, '123123');
    fireEvent.changeText(confirmPassword, '123123');
    fireEvent.press(submitButton);
  });

  const tree = toJSON();
  expect(tree).toMatchSnapshot();
});
