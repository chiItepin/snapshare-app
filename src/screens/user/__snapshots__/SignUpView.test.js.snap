// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpView renders and handles events 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
>
  <View>
    <View
      style={
        Array [
          undefined,
          undefined,
          undefined,
          undefined,
          Object {},
          Object {},
          Object {},
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "flex-start",
            "padding": 12,
          },
        ]
      }
    >
      <View
        style={
          Array [
            undefined,
            undefined,
            undefined,
            undefined,
            Object {},
            Object {},
            Object {},
            Object {
              "flex": 1,
              "paddingHorizontal": 5,
              "paddingVertical": 10,
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
              Object {
                "marginVertical": 12,
              },
            ]
          }
        >
          Signing Up
        </Text>
        <View
          style={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              Object {},
              Object {},
              Object {},
              Array [
                Object {},
                undefined,
              ],
            ]
          }
        >
          <Text
            color="#20303C"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "textAlign": "left",
                },
                undefined,
                Object {
                  "color": "#20303C",
                },
                undefined,
                undefined,
                Object {},
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "minHeight": 20,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
            testID="undefined.label"
          >
            Email
          </Text>
          <View
            centerV={true}
            row={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Array [
                  Object {},
                  Array [
                    Object {
                      "borderBottomWidth": 1,
                      "borderColor": "lightgrey",
                      "maxHeight": 200,
                      "paddingBottom": 10,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            <View
              row={true}
              style={
                Array [
                  undefined,
                  undefined,
                  Object {
                    "flex": 1,
                  },
                  undefined,
                  Object {},
                  Object {},
                  Object {
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <TextInput
                accessibilityState={
                  Object {
                    "disabled": false,
                  }
                }
                allowFontScaling={true}
                autoCapitalize="none"
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="Enter your email"
                placeholderTextColor="#6E7881"
                rejectResponderTermination={true}
                style={
                  Array [
                    Object {
                      "flexGrow": 1,
                      "paddingBottom": 0,
                      "paddingTop": 0,
                      "textAlign": "left",
                    },
                    Object {
                      "color": "#20303C",
                    },
                    Array [
                      Object {},
                      undefined,
                      undefined,
                    ],
                  ]
                }
                underlineColorAndroid="transparent"
                validate={
                  Array [
                    "required",
                    "email",
                  ]
                }
                validateOnChange={true}
                validationMessage={
                  Array [
                    "Email is required",
                    "Email is invalid",
                  ]
                }
                value="lorem@ipsum.com"
              />
            </View>
          </View>
          <View
            row={true}
            spread={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
                undefined,
              ]
            }
          >
            <Text
              red30={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "textAlign": "left",
                  },
                  undefined,
                  Object {
                    "color": "#FC3D2F",
                  },
                  undefined,
                  undefined,
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "minHeight": 20,
                    },
                    Object {
                      "marginBottom": 10,
                    },
                  ],
                ]
              }
              testID="undefined.validationMessage"
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              Object {},
              Object {},
              Object {},
              Array [
                Object {},
                undefined,
              ],
            ]
          }
        >
          <Text
            color="#20303C"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "textAlign": "left",
                },
                undefined,
                Object {
                  "color": "#20303C",
                },
                undefined,
                undefined,
                Object {},
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "minHeight": 20,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
            testID="undefined.label"
          >
            Password
          </Text>
          <View
            centerV={true}
            row={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Array [
                  Object {},
                  Array [
                    Object {
                      "borderBottomWidth": 1,
                      "borderColor": "lightgrey",
                      "maxHeight": 200,
                      "paddingBottom": 10,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            <View
              row={true}
              style={
                Array [
                  undefined,
                  undefined,
                  Object {
                    "flex": 1,
                  },
                  undefined,
                  Object {},
                  Object {},
                  Object {
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <TextInput
                accessibilityState={
                  Object {
                    "disabled": false,
                  }
                }
                allowFontScaling={true}
                maxLength={20}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="Enter your password"
                placeholderTextColor="#6E7881"
                rejectResponderTermination={true}
                secureTextEntry={true}
                style={
                  Array [
                    Object {
                      "flexGrow": 1,
                      "paddingBottom": 0,
                      "paddingTop": 0,
                      "textAlign": "left",
                    },
                    Object {
                      "color": "#20303C",
                    },
                    Array [
                      Object {},
                      undefined,
                      undefined,
                    ],
                  ]
                }
                underlineColorAndroid="transparent"
                value="123123"
              />
            </View>
          </View>
          <View
            row={true}
            spread={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
                undefined,
              ]
            }
          >
            <Text
              grey30={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "textAlign": "left",
                  },
                  undefined,
                  Object {
                    "color": "#6E7881",
                  },
                  undefined,
                  undefined,
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "flex": 1,
                      "textAlign": "right",
                    },
                    undefined,
                  ],
                ]
              }
            >
              6/20
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              Object {},
              Object {},
              Object {},
              Array [
                Object {},
                undefined,
              ],
            ]
          }
        >
          <Text
            color="#20303C"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "textAlign": "left",
                },
                undefined,
                Object {
                  "color": "#20303C",
                },
                undefined,
                undefined,
                Object {},
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "minHeight": 20,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
            testID="undefined.label"
          >
            Confirm Password
          </Text>
          <View
            centerV={true}
            row={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Array [
                  Object {},
                  Array [
                    Object {
                      "borderBottomWidth": 1,
                      "borderColor": "lightgrey",
                      "maxHeight": 200,
                      "paddingBottom": 10,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            <View
              row={true}
              style={
                Array [
                  undefined,
                  undefined,
                  Object {
                    "flex": 1,
                  },
                  undefined,
                  Object {},
                  Object {},
                  Object {
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <TextInput
                accessibilityState={
                  Object {
                    "disabled": false,
                  }
                }
                allowFontScaling={true}
                maxLength={20}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="Confirm your password"
                placeholderTextColor="#6E7881"
                rejectResponderTermination={true}
                secureTextEntry={true}
                style={
                  Array [
                    Object {
                      "flexGrow": 1,
                      "paddingBottom": 0,
                      "paddingTop": 0,
                      "textAlign": "left",
                    },
                    Object {
                      "color": "#20303C",
                    },
                    Array [
                      Object {},
                      undefined,
                      undefined,
                    ],
                  ]
                }
                underlineColorAndroid="transparent"
                value="123123"
              />
            </View>
          </View>
          <View
            row={true}
            spread={true}
            style={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
                undefined,
              ]
            }
          >
            <Text
              grey30={true}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "textAlign": "left",
                  },
                  undefined,
                  Object {
                    "color": "#6E7881",
                  },
                  undefined,
                  undefined,
                  Object {},
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "flex": 1,
                      "textAlign": "right",
                    },
                    undefined,
                  ],
                ]
              }
            >
              6/20
            </Text>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#5A48F5",
              "borderRadius": 6,
              "elevation": 2,
              "flexDirection": "row",
              "justifyContent": "center",
              "minWidth": 90,
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 9.5,
              "shadowColor": "#5A48F5",
              "shadowOffset": Object {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.35,
              "shadowRadius": 9.5,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "textAlign": "left",
                },
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "flex": 0,
                    "flexDirection": "row",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Object {
                    "color": "#FFFFFF",
                  },
                  Object {},
                  Object {},
                  undefined,
                ],
              ]
            }
            testID="undefined.label"
          >
            Sign Up
          </Text>
        </View>
        <View
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderRadius": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 12,
              "minWidth": undefined,
              "opacity": 1,
              "paddingHorizontal": undefined,
              "paddingVertical": undefined,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "textAlign": "left",
                },
                undefined,
                undefined,
                undefined,
                undefined,
                Object {},
                undefined,
                undefined,
                Object {
                  "textDecorationLine": "underline",
                },
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "flex": 0,
                    "flexDirection": "row",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Object {
                    "color": "#5A48F5",
                  },
                  Object {},
                  Object {
                    "fontFamily": "System",
                    "fontSize": 12,
                    "fontWeight": "700",
                    "lineHeight": 16,
                  },
                  Object {
                    "textDecorationLine": "none",
                  },
                ],
              ]
            }
            testID="undefined.label"
          >
            Already have an account
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
